### PR TITLE
Improving the init action test for Kafka

### PR DESCRIPTION
**Pull request by prince-cs**
For Kafka, adding the tests for submission of jobs and verifying instances to increase the coverage for overall scope. The test now runs kafka jobs with cruise-control and kafka-manager. Also, adding the test for cruise-control and kafka-manager instance verification.

### CODE REVIEW

File: kafka/test_kafka.py [hunk 1]
```diff
@@ -9,7 +10,15 @@
 class KafkaTestCase(DataprocTestCase):
     COMPONENT = 'kafka'
     INIT_ACTIONS = ['kafka/kafka.sh']
+    KAFKA_CRUISE_CONTROL_INIT_ACTION = ['kafka/kafka.sh', 'kafka/cruise-control.sh']
+    KAFKA_MANAGER_INIT_ACTION = ['kafka/kafka.sh', 'kafka/kafka-manager.sh']
     TEST_SCRIPT_FILE_NAME = 'validate.sh'
+    KAFKA_PYTHON_INIT_ACTION_PATH = 'gs://dataproc-kafka-test/kafka_python.sh'
+    KAFKA_PYTHON3_INIT_ACTION_PATH = 'gs://dataproc-kafka-test/kafka_python3.sh'
+    PYTHON2_VERSION = 'python2.7'
+    PYTHON3_VERSION = 'python3.10'
```
Comment (functicons) [inline kafka/test_kafka.py:19]: 3.10 might not be the right version for all Dataproc versions. I think we should use `python3` here.

File: kafka/verify_kafka_running.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+def test_produce_and_consume_messages():
+    cluster_name = get_cluster_name()
+    kafka_runner = KafkaTestRunner(cluster_name)
+    kafka_runner.create_test_topic()
+    kafka_runner.publish_messages()
+    kafka_runner.consume_messages()
+
+
+def main():
+    time.sleep(10)
```
Comment (functicons) [inline kafka/verify_kafka_running.py:153]: Is it to make sure Kafka is up and running? I don't think we should add it in the test, because users won't do the same. If wait is really necessary, it should be in the Kafka init action.

File: kafka/test_kafka.py [hunk 3]
```diff
@@ -23,6 +32,22 @@ def __run_test_script(self, name):
         self.assert_instance_command(
             name, "bash {}".format(self.TEST_SCRIPT_FILE_NAME))
 
+    def __submit_pyspark_job(self, cluster_name):
+        if self.getImageVersion() >= pkg_resources.parse_version("2.1"):
+            verify_script = self.KAFKA_VERIFY_SCRIPT_PYTHON3
+            python_version = self.PYTHON3_VERSION
+        else:
+            verify_script = self.KAFKA_VERIFY_SCRIPT_PYTHON2
```
Comment (functicons) [inline kafka/test_kafka.py:40]: verify_kafka_running_py3.py and verify_kafka_running.py have a lot in common. I think we should merge then, and pass in parameters or detect different image versions inside the script.

File: kafka/test_kafka.py [hunk 4]
```diff
@@ -23,6 +32,22 @@ def __run_test_script(self, name):
...
+            python_version = self.PYTHON2_VERSION
```
Comment (functicons) [inline kafka/test_kafka.py:41]: Do you know what happens if we use Python3 in 2.0?

File: kafka/test_kafka.py [hunk 5]
```diff
@@ -9,7 +10,15 @@
 class KafkaTestCase(DataprocTestCase):
     COMPONENT = 'kafka'
     INIT_ACTIONS = ['kafka/kafka.sh']
+    KAFKA_CRUISE_CONTROL_INIT_ACTION = ['kafka/kafka.sh', 'kafka/cruise-control.sh']
+    KAFKA_MANAGER_INIT_ACTION = ['kafka/kafka.sh', 'kafka/kafka-manager.sh']
     TEST_SCRIPT_FILE_NAME = 'validate.sh'
+    KAFKA_PYTHON_INIT_ACTION_PATH = 'gs://dataproc-kafka-test/kafka_python.sh'
```
Comment (functicons) [inline kafka/test_kafka.py:16]: I think we should merge the 2 init actions into one, pass parameters or detect the env inside the init action.

**Review by functicons — CHANGES_REQUESTED** [review]

File: kafka/test_kafka.py (hunk 1 above)
Comment (prince-cs) [reply kafka/test_kafka.py:19]: done

File: kafka/test_kafka.py (hunk 3 above)
Comment (prince-cs) [reply kafka/test_kafka.py]: based on the dataproc image version both the scripts are merged and tested locally.

File: kafka/test_kafka.py (hunk 5 above)
Comment (prince-cs) [reply kafka/test_kafka.py:16]: merging them would require to change the script at the location at which it is stored as it is custom made init action script for kafka and while running with the repo it is not able to fetch this file. So,

Comment (functicons) [reply kafka/test_kafka.py:16]: I don't quite understand why it needs to be in this GCS bucket. Cannot it simply use the one from this repo? Just like the kafka.sh itself.

**Review by functicons — CHANGES_REQUESTED** [review]

File: kafka/test_kafka.py (hunk 5 above)
Comment (prince-cs) [reply kafka/test_kafka.py:16]: I tried that but it started throwing error saying file not found.

Comment (functicons) [reply kafka/test_kafka.py:16]: It should be the same as the kafka init action itself. Or you can extend the kafka init action with parameters to allow you do the work, this will bypass the problem and mabye some users also like it.

**Review by functicons — CHANGES_REQUESTED** [review]

File: kafka/test_kafka.py (hunk 5 above)
Comment (prince-cs) [reply kafka/test_kafka.py:16]: Passing the parameter and based on the dataproc image version being used for the init action, installing the kafka-python package accordingly with kafka init action itself.

File: kafka/kafka.sh [hunk 6]
```diff
@@ -25,6 +25,7 @@ readonly ROLE="$(/usr/share/google/get_metadata_value attributes/dataproc-role)"
 readonly RUN_ON_MASTER="$(/usr/share/google/get_metadata_value attributes/run-on-master || echo false)"
 readonly KAFKA_ENABLE_JMX="$(/usr/share/google/get_metadata_value attributes/kafka-enable-jmx || echo false)"
 readonly KAFKA_JMX_PORT="$(/usr/share/google/get_metadata_value attributes/kafka-jmx-port || echo 9999)"
+readonly DATAPROC_IMAGE="$(/usr/share/google/get_metadata_value attributes/image-version)"
```
Comment (functicons) [inline kafka/kafka.sh:28]: Why is this added? It's not consistent with how other init actions get the image version.

File: kafka/kafka.sh [hunk 7]
```diff
@@ -168,8 +174,25 @@ function install_and_configure_kafka_server() {
...
+      apt-get install -y python-pip
+    fi
+    pip2 install "${KAFKA_PYTHON_PACKAGE}" || { sleep 10; pip2 install "${KAFKA_PYTHON_PACKAGE}"; } || { sleep 10; pip install "${KAFKA_PYTHON_PACKAGE}"; }
+  fi
+}
+
 function main() {
+  recv_keys || err 'Unable to receive keys.'
   update_apt_get || err 'Unable to update packages lists.'
+  install_kafka_python_package
```
Comment (functicons) [inline kafka/kafka.sh:195]: This new functionality should be guarded by a new meta data, e.g., install-kafka-python with default false.

**Review by functicons — CHANGES_REQUESTED** [review]

File: kafka/kafka.sh (hunk 6 above)
Comment (prince-cs) [reply kafka/kafka.sh:28]: Okay got it. We have the DATAPROC_IMAGE_VERSION for this. No need to define it under the metadata.

File: kafka/kafka.sh (hunk 7 above)
Comment (prince-cs) [reply kafka/kafka.sh:195]: Understood.

File: kafka/kafka.sh [hunk 8]
```diff
@@ -168,8 +174,27 @@ function install_and_configure_kafka_server() {
   wait_for_kafka
 }
 
+function install_kafka_python_package() {
+  KAFKA_PYTHON_PACKAGE="kafka-python==2.0.2"
+  if [[ "${INSTALL_KAFKA_PYTHON}" == "true" ]]; then
```
Comment (functicons) [inline kafka/kafka.sh:179]: Prefer shortcut:
```
if [[ "${INSTALL_KAFKA_PYTHON}" != "true" ]]; then
  return
fi
```

File: kafka/kafka.sh [hunk 9]
```diff
@@ -25,6 +25,7 @@ readonly ROLE="$(/usr/share/google/get_metadata_value attributes/dataproc-role)"
 readonly RUN_ON_MASTER="$(/usr/share/google/get_metadata_value attributes/run-on-master || echo false)"
 readonly KAFKA_ENABLE_JMX="$(/usr/share/google/get_metadata_value attributes/kafka-enable-jmx || echo false)"
 readonly KAFKA_JMX_PORT="$(/usr/share/google/get_metadata_value attributes/kafka-jmx-port || echo 9999)"
+readonly INSTALL_KAFKA_PYTHON="$(/usr/share/google/get_metadata_value attributes/install-kafka-python || echo false)"
```
Comment (functicons) [inline kafka/kafka.sh:28]: The new metadata should be documented in the README file.

**Review by functicons — CHANGES_REQUESTED** [review]

File: kafka/kafka.sh (hunk 8 above)
Comment (prince-cs) [reply kafka/kafka.sh]: Done

File: kafka/kafka.sh (hunk 9 above)
Comment (prince-cs) [reply kafka/kafka.sh:28]: Done.

**Review by functicons — APPROVED** [review]
LGTM
